drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,878 @@
+{
+	struct smb_reg *const inst = HAL_I2C_INSTANCE(dev);
+
+	/* Enable writing to SCL_LVL/SDA_LVL bit in SMBnCTL3 */
+	inst->SMBCTL4 |= BIT(NPCX_SMBCTL4_LVL_WE);
+	/*
+	 * Release SCL bus. Then it might be still driven by module itself or
+	 * slave device.
+	 */
+	inst->SMBCTL3 |= BIT(NPCX_SMBCTL3_SCL_LVL) | BIT(NPCX_SMBCTL3_SDA_LVL);
+	/* Disable writing to them */
+	inst->SMBCTL4 &= ~BIT(NPCX_SMBCTL4_LVL_WE);
+}
+
+/* I2C controller inline functions operate in 'FIFO' mode */
+static inline void i2c_ctrl_write(const struct device *dev, uint8_t data)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	inst_fifo->SMBSDA = data;
+}
+
+static inline uint8_t i2c_ctrl_read(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	return inst_fifo->SMBSDA;
+}
+
+static inline void i2c_ctrl_start(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	inst_fifo->SMBCTL1 |= BIT(NPCX_SMBCTL1_START);
+}
+
+static inline void i2c_ctrl_stop(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	inst_fifo->SMBCTL1 |= BIT(NPCX_SMBCTL1_STOP);
+}
+
+static inline int i2c_ctrl_bus_busy(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	return IS_BIT_SET(inst_fifo->SMBCST, NPCX_SMBCST_BB);
+}
+
+static inline int i2c_ctrl_tx_fifo_avail(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	return NPCX_I2C_FIFO_MAX_SIZE - (inst_fifo->SMBTXF_STS & 0x3f);
+}
+
+static inline int i2c_ctrl_rx_fifo_occupied(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	return inst_fifo->SMBRXF_STS & 0x3f;
+}
+
+static inline void i2c_ctrl_rx_fifo_setup_threshold_nack(
+		const struct device *dev, int threshold, int last)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+	uint8_t value = MIN(threshold, NPCX_I2C_FIFO_MAX_SIZE);
+
+	SET_FIELD(inst_fifo->SMBRXF_CTL, NPCX_SMBRXF_CTL_RX_THR, value);
+
+	/*
+	 * Is it last received transaction? If so, set LAST bit. Then the
+	 * hardware will generate NACK automatically when receiving last byte.
+	 */
+	if (last && (value == threshold))
+		inst_fifo->SMBRXF_CTL |= BIT(NPCX_SMBRXF_CTL_LAST);
+}
+
+static inline void i2c_ctrl_clear_fifo_status(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+
+	inst_fifo->SMBFIF_CTS |= BIT(NPCX_SMBFIF_CTS_CLR_FIFO);
+}
+
+
+static inline void i2c_ctrl_irq_enable(const struct device *dev, int enable)
+{
+	const struct i2c_ctrl_config *const config = DRV_CONFIG(dev);
+
+	if (enable)
+		irq_enable(config->irq);
+	else
+		irq_disable(config->irq);
+}
+/*
+ * I2C local functions operate in 'Normal' mode. Please notice we should change
+ * bank back to FIFO mode when leaving function in case unexpected reigster
```

Comment:
> What is an example of unexptected register access? If all of these function set the mode they need before accessing the HAL, then there shouldn't be anything unexpected right?
> Changing it back to FIFO every time seems rather confusing
Since the partial register of Normal and FIFO registers are overlapped, we need to check bank selection before accessing them. Yes, we can set the mode in every i2c function. But besides i2c_ctrl_bank_select(), the other normal mode functions are only called once during initialization. That's why we stay in FIFO mode most of the time for effective concern.

Reply:
I will refactor it and add more comments later.

Reply:
Done.